File: tools/js/commit_integrity_validator.js
/**
 * Commit 诚信验证工具 - 对比 Commit Message 与物理代码变更的一致性
 * 
 * 功能说明：
 * - 解析拟提交的 Commit Message (WHAT/WHY/HOW 结构)
 * - 获取当前 Git 暂存区 (staged) 或未暂存 (unstaged) 的文件变更
 * - 对比 HOW 字段中提到的文件与实际变更文件是否匹配
 * - 识别遗漏的文件或未在 HOW 中说明的变更文件
 * - 输出诚信分 (Integrity Score) 和改进建议
 * 
 * 使用方法：
 *     node tools/js/commit_integrity_validator.js --message "COMMIT_MESSAGE" [--staged] [--unstaged] [--format json]
 * 
 * 参数说明：
 *     --message TEXT      要验证的 Commit Message（必需）
 *     --staged            检查已暂存(staged)的变更（默认）
 *     --unstaged          检查未暂存(unstaged)的变更
 *     --format FORMAT     输出格式：text, json（默认：text）
 * 
 * 版本信息：
 *     版本：1.1.0
 *     更新日期：2026-04-22
 */

const { execSync } = require('child_process');

/**
 * 获取 Git 变更的文件列表
 */
function getGitDiffFiles(staged = true) {
    const cmd = staged ? 'git diff --name-only --staged' : 'git diff --name-only';
    try {
        const output = execSync(cmd, { encoding: 'utf8', timeout: 5000 });
        return new Set(output.split('\n').map(line => line.trim()).filter(line => line.length > 0));
    } catch (error) {
        return new Set();
    }
}

/**
 * 从 Commit Message 的 HOW 字段中提取提到的文件路径
 */
function parseHowFiles(message) {
    let howSection = "";
    // 查找 HOW: 之后的所有内容, 直到下一个大写关键字 (如 WHY, WHAT) 或结束
    const howMatch = message.match(/HOW:\s*([\s\S]*?)(?=\s*[A-Z]{3,}:|$)/i);
    if (howMatch) {
        howSection = howMatch[1];
    }

    // 核心优化：移除所有括号及其内容，防止误读描述 (如 "Node.js", "AI")
    const cleanHow = howSection.replace(/\(.*?\)/g, '');

    const foundPaths = new Set();
    // 路径匹配正则: 匹配包含斜杠的文件名, 或者带常见扩展名的文件名
    const patterns = [
        /\b[a-zA-Z0-9_\-\./]+\.[a-zA-Z0-9]{1,10}\b/g, // 标准路径或带扩展名文件
        /\.[a-zA-Z0-9_\-]+\b/g                        // 隐藏文件如 .gitignore
    ];

    patterns.forEach(pattern => {
        const matches = cleanHow.match(pattern);
        if (matches) {
            matches.forEach(m => {
                // 清理标点
                let p = m.replace(/[.,:; "']+$/, '').replace(/^[.,:; "']+/, '');
                // 统一处理 ./ 前缀
                if (p.startsWith('./')) {
                    p = p.substring(2);
                }
                if (p) {
                    foundPaths.add(p);
                }
            });
        }
    });

    return foundPaths;
}

/**
 * 计算诚信分并生成报告
 */
function calculateIntegrity(msgFiles, actualFiles) {
    const matched = new Set([...msgFiles].filter(x => actualFiles.has(x)));
    const overReported = new Set([...msgFiles].filter(x => !actualFiles.has(x)));
    const underReported = new Set([...actualFiles].filter(x => !msgFiles.has(x)));

    let score = 100;
    const underReportedPenalty = underReported.size * 15;
    const overReportedPenalty = overReported.size * 5;

    score -= (underReportedPenalty + overReportedPenalty);
    score = Math.max(0, score);

    const suggestions = [];
    if (underReported.size > 0) {
        suggestions.push(`⚠️ 以下文件已修改但未在 HOW 中说明: ${Array.from(underReported).join(', ')}`);
    }
    if (overReported.size > 0) {
        suggestions.push(`ℹ️ HOW 中提到的以下文件似乎未发生变更: ${Array.from(overReported).join(', ')}`);
    }

    let status = "NEEDS_IMPROVEMENT";
    if (score >= 90) status = "EXCELLENT";
    else if (score >= 80) status = "GOOD";

    return {
        score,
        status,
        matched_count: matched.size,
        under_reported: Array.from(underReported),
        over_reported: Array.from(overReported),
        suggestions
    };
}

function main() {
    const startTime = Date.now();
    const args = process.argv.slice(2);
    let message = "";
    let staged = true;
    let format = "text";

    for (let i = 0; i < args.length; i++) {
        if (args[i] === '--message' && args[i + 1]) {
            message = args[i + 1];
            i++;
        } else if (args[i] === '--unstaged') {
            staged = false;
        } else if (args[i] === '--staged') {
            staged = true;
        } else if (args[i] === '--format' && args[i + 1]) {
            format = args[i + 1];
            i++;
        }
    }

    if (!message) {
        if (format === 'json') {
            console.log(JSON.stringify({ error: "No message provided" }, null, 2));
        } else {
            console.error("错误: 必须通过 --message 提供 Commit 信息");
        }
        process.exit(1);
    }

    const actualFiles = getGitDiffFiles(staged);
    const msgFiles = parseHowFiles(message);
    const reportData = calculateIntegrity(msgFiles, actualFiles);

    const elapsedTime = ((Date.now() - startTime) / 1000).toFixed(4);
    const result = {
        data: reportData,
        metadata: {
            elapsed_seconds: parseFloat(elapsedTime),
            version: '1.1.0'
        }
    };

    if (format === 'json') {
        console.log(JSON.stringify(result, null, 2));
    } else {
        console.log("--- Commit 诚信审计报告 ---");
        console.log(`得分: ${result.data.score} [${result.data.status}]`);
        console.log(`匹配文件数: ${result.data.matched_count}`);

        if (result.data.suggestions.length > 0) {
            console.log("\n建议:");
            result.data.suggestions.forEach(s => console.log(`  ${s}`));
        } else {
            console.log("\n✅ 意图与物理变更完美匹配！");
        }
    }
}

if (require.main === module) {
    main();
}

File: tools/py/commit_integrity_validator.py
"""
Commit 诚信验证工具 - 对比 Commit Message 与物理代码变更的一致性

功能说明：
- 解析擬提交的 Commit Message (WHAT/WHY/HOW 结构)
- 获取当前 Git 暂存区 (staged) 或未暂存 (unstaged) 的文件变更
- 对比 HOW 字段中提到的文件与实际变更文件是否匹配
- 识别遗漏的文件或未在 HOW 中说明的变更文件
- 输出诚信分 (Integrity Score) 和改进建议

使用方法：
    python tools/py/commit_integrity_validator.py --message "COMMIT_MESSAGE" [--staged]

参数说明：
    --message TEXT      要验证的 Commit Message（必需）
    --staged            检查已暂存(staged)的变更（默认）
    --unstaged          检查未暂存(unstaged)的变更
    --format FORMAT     输出格式：text, json（默认：text）

输出格式：
    text 格式：
    - Commit 诚信审计报告
    - 得分：[分数] [状态]
    - 匹配文件数：N
    - 建议列表（如有）

    json 格式：
    {
      "score": 85,
      "status": "EXCELLENT",
      "matched_count": 3,
      "under_reported": [...],
      "over_reported": [...],
      "suggestions": [...]
    }

版本信息：
    版本：1.0.0
    更新日期：2026-04-07
"""

import re
import json
import subprocess
import argparse
from typing import List, Dict, Set

def parse_how_files(message: str) -> Set[str]:
    """
    从 Commit Message 的 HOW 字段中提取提到的文件路径 (生产级解析器 v3)
    
    解析逻辑:
    1. 提取 HOW: 之后的所有文本
    2. 忽略括号 (...) 中的内容，防止误将描述文字识别为路径
    3. 使用更精准的正则匹配完整路径
    """
    how_section = ""
    # 查找 HOW: 之后的所有内容, 直到下一个大写关键字 (如 WHY, WHAT) 或结束
    how_match = re.search(r'HOW:\s*(.*?)(?=\s*[A-Z]{3,}:|$)', message, re.DOTALL)
    if how_match:
        how_section = how_match.group(1)
    
    # 核心优化：移除所有括号及其内容，防止误读描述 (如 "Node.js", "AI")
    clean_how = re.sub(r'\(.*?\)', '', how_section)
    
    # 路径匹配正则: 匹配包含斜杠的文件名, 或者带常见扩展名的文件名
    # 使用 \b 单词边界防止匹配到类似 v1.4 的版本号
    patterns = [
        r'\b[a-zA-Z0-9_\-\./]+\.[a-zA-Z0-9]{1,10}\b', # 标准路径或带扩展名文件
        r'(?<![a-zA-Z0-9_\-/])\.[a-zA-Z0-9_\-]+(?:\.[a-zA-Z0-9_\-]+)?\b'  # 隐藏文件如 .gitignore
    ]
    
    found_paths = set()
    for pattern in patterns:
        matches = re.findall(pattern, clean_how)
        for m in matches:
            # 清理标点
            p = m.strip(' "\'').rstrip('.,:;')
            # 统一处理 ./ 前缀
            if p.startswith('./'):
                p = p[2:]
            if p:
                found_paths.add(p)
    
    return found_paths

def get_git_diff_files(staged: bool = True) -> Set[str]:
    """获取 Git 变更的文件列表 (包含超时保护)"""
    cmd = ["git", "diff", "--name-only"]
    if staged:
        cmd.append("--staged")
    
    try:
        # 增加 5 秒超时保护, 防止在巨型仓库中挂死
        result = subprocess.run(cmd, capture_output=True, text=True, check=True, timeout=5)
        return {line.strip() for line in result.stdout.splitlines() if line.strip()}
    except (subprocess.CalledProcessError, subprocess.TimeoutExpired):
        return set()

def calculate_integrity(msg_files: Set[str], actual_files: Set[str]) -> Dict:
    """计算诚信分并生成报告"""
    # 1. 匹配到的文件
    matched = msg_files.intersection(actual_files)
    
    # 2. Commit 提到了但实际没改的文件 (虚报)
    over_reported = msg_files - actual_files
    
    # 3. 实际改了但 Commit 没提的文件 (漏报)
    under_reported = actual_files - msg_files
    
    # 计算分数 (基础分 100)
    score = 100
    
    # 惩罚项
    under_reported_penalty = len(under_reported) * 15 # 漏报严重，扣分多
    over_reported_penalty = len(over_reported) * 5    # 虚报较轻，可能只是路径写错
    
    score -= (under_reported_penalty + over_reported_penalty)
    score = max(0, score) # 不低于0分
    
    # 生成建议
    suggestions = []
    if under_reported:
        suggestions.append(f"⚠️ 以下文件已修改但未在 HOW 中说明: {', '.join(under_reported)}")
    if over_reported:
        suggestions.append(f"ℹ️ HOW 中提到的以下文件似乎未发生变更: {', '.join(over_reported)}")
    
    if score >= 90:
        status = "EXCELLENT"
    elif score >= 80:
        status = "GOOD"
    else:
        status = "NEEDS_IMPROVEMENT"
        
    return {
        "score": score,
        "status": status,
        "matched_count": len(matched),
        "under_reported": list(under_reported),
        "over_reported": list(over_reported),
        "suggestions": suggestions
    }

def main():
    parser = argparse.ArgumentParser(description='Commit Integrity Validator')
    parser.add_argument('--message', required=True, help='Commit message to validate')
    parser.add_argument('--staged', action='store_true', default=True, help='Check staged changes (default)')
    parser.add_argument('--unstaged', action='store_false', dest='staged', help='Check unstaged changes')
    parser.add_argument('--format', choices=['text', 'json'], default='text', help='Output format')
    
    args = parser.parse_args()
    
    actual_files = get_git_diff_files(args.staged)
    msg_files = parse_how_files(args.message)
    
    report = calculate_integrity(msg_files, actual_files)
    
    if args.format == 'json':
        print(json.dumps(report, indent=2, ensure_ascii=False))
    else:
        print(f"--- Commit 诚信审计报告 ---")
        print(f"得分: {report['score']} [{report['status']}]")
        print(f"匹配文件数: {report['matched_count']}")
        
        if report['suggestions']:
            print("\n建议:")
            for s in report['suggestions']:
                print(f"  {s}")
        else:
            print("\n✅ 意图与物理变更完美匹配！")

if __name__ == "__main__":
    main()
